### PR TITLE
vnc namespaces

### PR DESCRIPTION
**Pull request by cdshann**
This PR makes the VNC api namespace aware, and also moves the vnc recordings onto the host where the vm is running. 

Updates #426 .

### CODE REVIEW

File: src/minimega/vnc_cli.go [hunk 1]
```diff
@@ -27,68 +28,101 @@ If playback is selected, the specified file (created using vnc record) will be
...
 		HelpShort: "reset VNC state",
 		HelpLong: `
 Resets the state for VNC recordings. See "help vnc" for more information.`,
 		Patterns: []string{
 			"clear vnc",
 		},
-		Call: wrapSimpleCLI(func(_ *minicli.Command, _ *minicli.Response) error {
-			return vncClear()
-		}),
+		Call: wrapVMTargetCLI(cliVNCClear),
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Change to `wrapBroadcastCLI`.

File: src/minimega/vnc_cli.go [hunk 2]
```diff
@@ -27,68 +28,101 @@ If playback is selected, the specified file (created using vnc record) will be
...
+List all running vnc playback/recording instances. See "help vnc" for more information.`,
+		Patterns: []string{
+			"vnc",
+		},
+		Call: wrapBroadcastCLI(vncList),
 	},
 }
 
+// List all active recordings and playbacks
+func vncList(c *minicli.Command, resp *minicli.Response) error {
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Rename to `cliVNCList`. Should only list if it matches the active namespace.

File: src/minimega/vnc.go [hunk 3]
```diff
@@ -61,37 +60,21 @@ func init() {
...
+		return nil, vmNotFound(name)
 	}
 
-	rhost := fmt.Sprintf("%v:%v", host, vm.VNCPort)
+	rhost := fmt.Sprintf("%v:%v", name, vm.VNCPort)
 
 	c := &vncClient{
 		Rhost: rhost,
-		Host:  host,
+		Host:  hostname,
```
Comment (jcrussell) [inline src/minimega/vnc.go]: Might be worth adding a reference to the VM to the `vncClient` struct. Then you don't need to store Name or ID.

File: src/minimega/vnc_cli.go [hunk 4]
```diff
@@ -103,36 +137,37 @@ func cliVNC(c *minicli.Command, resp *minicli.Response) error {
...
-		return vncPlaybackKB(host, vm, fname)
+		return vncPlaybackKB(vm, fname)
 	}
+	return nil
+}
 
-	// List all active recordings and playbacks
-	resp.Header = []string{"host", "name", "id", "type", "filename"}
-	resp.Tabular = [][]string{}
+func cliVNCClear(c *minicli.Command, resp *minicli.Response) error {
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Should we do something akin to `cliVNCClear` when a VM is flushed? What cleans up these maps?

File: src/minimega/tunnel.go [hunk 5]
```diff
@@ -17,79 +17,70 @@ import (
...
-					if !strings.Contains(err.Error(), "closed network connection") && err != io.EOF {
-						log.Errorln(err)
-					}
-					break
-				}
-
-				err = websocket.Message.Send(ws, rbuf[:n])
-				if err != nil {
+		rbuf := make([]byte, VNC_WS_BUF)
+		for {
```
Comment (jcrussell) [inline src/minimega/tunnel.go]: This *seems* unnecessary... we're already copying from remote to ws above. Should figure this out.

File: src/minimega/vnc.go [hunk 6]
```diff
@@ -26,9 +25,8 @@ var (
 )
 
 type vncClient struct {
-	Host  string
-	Name  string
-	ID    int
+	Vm    *KvmVM
+	Ns    string
```
Comment (jcrussell) [inline src/minimega/vnc.go]: `Vm` -> `VM`. `Ns` -> `ID // ID is namespace:name`.

File: src/minimega/vnc_cli.go [hunk 7]
```diff
@@ -27,72 +28,106 @@ If playback is selected, the specified file (created using vnc record) will be
...
 		Patterns: []string{
-			"vnc",
-			"vnc <kb,fb> <record,> <host> <vm id or name> <filename>",
-			"vnc <kb,fb> <norecord,> <host> <vm id or name>",
-			"vnc <playback,> <host> <vm id or name> <filename>",
-			"vnc <noplayback,> <host> <vm id or name>",
+			"vnc <kb,fb> <record,> <vm name> <filename>",
+			"vnc <kb,fb> <norecord,> <vm name>",
+			"vnc <playback,> <vm name> <filename>",
+			"vnc <noplayback,> <vm name>",
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: While we're changing things... should we swap the order of `<kb,fb>` and `<record>`?

File: src/minimega/vnc_cli.go [hunk 8]
```diff
@@ -27,72 +28,106 @@ If playback is selected, the specified file (created using vnc record) will be
...
-		host = hostname
+	vm, err := vms.FindKvmVM(c.StringArgs["vm"])
+	if err != nil {
+		return fmt.Errorf("vm %s not found", c.StringArgs["vm"])
 	}
 
 	if c.BoolArgs["record"] && c.BoolArgs["kb"] {
 		// Starting keyboard recording
-		return vncRecordKB(host, vm, fname)
+		return vncRecordKB(vm.Name, fname)
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: We already have a reference to the VM -- we should pass it in here and then we don't have to find it again later in NewVNCClient.

File: src/minimega/vnc_cli.go [hunk 9]
```diff
@@ -27,72 +28,106 @@ If playback is selected, the specified file (created using vnc record) will be
...
 	} else if c.BoolArgs["record"] && c.BoolArgs["fb"] {
 		// Starting framebuffer recording
-		return vncRecordFB(host, vm, fname)
+		return vncRecordFB(vm.Name, fname)
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Same as above.

File: src/minimega/vnc_cli.go [hunk 10]
```diff
@@ -27,72 +28,106 @@ If playback is selected, the specified file (created using vnc record) will be
...
-			for k, v := range vncKBRecording {
-				if v.Matches(host, vm) {
-					client = v.vncClient
-					delete(vncKBRecording, k)
-					break
-				}
+			err = fmt.Errorf("kb recording %v not found", vm.Name)
+			if v, ok := vncKBRecording[ns]; ok {
+				client = v.vncClient
+				delete(vncKBRecording, ns)
```
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Is this the right place to delete from `vncKBRecording`? Need to think about this.

File: src/minimega/vnc_cli.go (hunk 4 above)
Comment (jcrussell) [inline src/minimega/vnc_cli.go]: Ah, now I see what you did here. I would move the current `cliVNCClear` code into `vncClear` and change the `clear vnc` handler to:

```
        Call: wrapBroadcastCLI(func(_ *minicli.Command, _ *minicli.Response) error {
            vncClear()
            return nil
        }),
```

File: src/minimega/kvm.go [hunk 11]
```diff
@@ -428,6 +429,7 @@ func (vm *KvmVM) connectVNC() error {
 	// Keep track of shim so that we can close it later
 	vm.vncShim = l
 	vm.VNCPort = l.Addr().(*net.TCPAddr).Port
+	ns := fmt.Sprintf("%v:%v", GetNamespaceName(), vm.Name)
```
Comment (jcrussell) [inline src/minimega/kvm.go]: This should be the VM's name, not the active namespace.

**Comment by jcrussell** [comment]
We're going to punt on the issue of adding a sync.RWLock for the vnc* maps. Will be addressed in #513.

**Comment by djfritz** [comment]
Will merge after the branch gets fixed.

**Comment by cdshann** [comment]
Don't merge me yet.

**Pull request merged**